File: sdks/go/chio-go/invariants/receipt.go
package invariants

type ReceiptVerification struct {
	Decision           string `json:"decision"`
	Authorized         bool   `json:"authorized"`
	BoundaryClass      string `json:"boundary_class"`
	Ok                 bool   `json:"ok"`
	ParameterHashValid bool   `json:"parameter_hash_valid"`
	ReceiptIDValid     bool   `json:"receipt_id_valid"`
	ReceiptKind        string `json:"receipt_kind"`
	Result             string `json:"result"`
	SignerKeyHex       string `json:"signer_key_hex"`
	SignerTrusted      bool   `json:"signer_trusted"`
	SignatureValid     bool   `json:"signature_valid"`
	TrustLevel         string `json:"trust_level"`
}

func ParseReceiptJSON(input string) (map[string]any, error) {
	value, err := ParseJSONText(input)
	if err != nil {
		return nil, err
	}
	receipt, ok := value.(map[string]any)
	if !ok {
		return nil, newInvariantError("json", "receipt must be a JSON object")
	}
	return receipt, nil
}

func ReceiptBodyCanonicalJSON(receipt map[string]any) (string, error) {
	body := make(map[string]any, len(receipt))
	for key, value := range receipt {
		if key == "algorithm" || key == "signature" {
			continue
		}
		body[key] = value
	}
	return CanonicalizeJSON(body)
}

func ReceiptSigningBodyCanonicalJSON(receipt map[string]any) (string, error) {
	receiptID, err := stringField(receipt, "id")
	if err != nil {
		return "", err
	}
	idInput, err := receiptIDInput(receipt)
	if err != nil {
		return "", err
	}
	return CanonicalizeJSON(map[string]any{
		"id":   receiptID,
		"body": idInput,
	})
}

func VerifyReceipt(receipt map[string]any) (ReceiptVerification, error) {
	return VerifyReceiptWithTrustedSigners(receipt, nil)
}

func VerifyReceiptWithTrustedSigners(
	receipt map[string]any,
	trustedSigners []string,
) (ReceiptVerification, error) {
	body, err := ReceiptSigningBodyCanonicalJSON(receipt)
	if err != nil {
		return ReceiptVerification{}, err
	}
	kernelKey, err := stringField(receipt, "kernel_key")
	if err != nil {
		return ReceiptVerification{}, err
	}
	signature, err := stringField(receipt, "signature")
	if err != nil {
		return ReceiptVerification{}, err
	}
	signatureValid, err := VerifyUTF8MessageEd25519(body, kernelKey, signature)
	if err != nil {
		return ReceiptVerification{}, err
	}
	receiptID, err := stringField(receipt, "id")
	if err != nil {
		return ReceiptVerification{}, err
	}
	computedReceiptID, err := contentAddressedReceiptID(receipt)
	if err != nil {
		return ReceiptVerification{}, err
	}
	receiptIDValid := receiptID == computedReceiptID
	if !receiptIDValid {
		signatureValid = false
	}
	action, err := mapField(receipt, "action")
	if err != nil {
		return ReceiptVerification{}, err
	}
	parameters, ok := action["parameters"]
	if !ok {
		return ReceiptVerification{}, newInvariantError("receipt", "missing field parameters")
	}
	renderedParameters, err := CanonicalizeJSON(parameters)
	if err != nil {
		return ReceiptVerification{}, err
	}
	expectedParameterHash := SHA256HexUTF8(renderedParameters)
	actualParameterHash, err := stringField(action, "parameter_hash")
	if err != nil {
		return ReceiptVerification{}, err
	}
	verdict := "none"
	if decision, ok := receipt["decision"]; ok && decision != nil {
		decisionMap, ok := decision.(map[string]any)
		if !ok {
			return ReceiptVerification{}, newInvariantError("receipt", "field decision must be an object")
		}
		var err error
		verdict, err = stringField(decisionMap, "verdict")
		if err != nil {
			return ReceiptVerification{}, err
		}
	}
	receiptKind, boundaryClass := receiptSemantics(receipt)
	trustLevel, _ := receipt["trust_level"].(string)
	if !semanticallySignable(receipt, verdict) {
		signatureValid = false
	}
	semanticAuthorized := receiptKind == "mediated_decision" && boundaryClass == "prevent" && verdict == "allow"
	signerTrusted := false
	if len(trustedSigners) > 0 {
		for _, signer := range trustedSigners {
			if PublicKeyHexMatches(signer, kernelKey) {
				signerTrusted = true
				break
			}
		}
	}
	parameterHashValid := actualParameterHash == expectedParameterHash
	authorized := semanticAuthorized && signatureValid && parameterHashValid && receiptIDValid && signerTrusted
	return ReceiptVerification{
		Decision:           verdict,
		Authorized:         authorized,
		BoundaryClass:      boundaryClass,
		Ok:                 signatureValid && parameterHashValid && receiptIDValid && signerTrusted,
		ParameterHashValid: parameterHashValid,
		ReceiptIDValid:     receiptIDValid,
		ReceiptKind:        receiptKind,
		Result:             resultLabel(receiptKind, boundaryClass, verdict),
		SignerKeyHex:       kernelKey,
		SignerTrusted:      signerTrusted,
		SignatureValid:     signatureValid,
		TrustLevel:         trustLevel,
	}, nil
}

func VerifyReceiptJSON(input string) (ReceiptVerification, error) {
	receipt, err := ParseReceiptJSON(input)
	if err != nil {
		return ReceiptVerification{}, err
	}
	return VerifyReceipt(receipt)
}

func contentAddressedReceiptID(receipt map[string]any) (string, error) {
	input, err := receiptIDInput(receipt)
	if err != nil {
		return "", err
	}
	canonical, err := CanonicalizeJSON(input)
	if err != nil {
		return "", err
	}
	return SHA256HexUTF8(canonical), nil
}

func receiptIDInput(receipt map[string]any) (map[string]any, error) {
	input := map[string]any{}
	for _, key := range []string{
		"timestamp",
		"capability_id",
		"tool_server",
		"tool_name",
		"action",
		"receipt_kind",
		"boundary_class",
		"tool_origin",
		"redaction_mode",
		"content_hash",
		"policy_hash",
		"kernel_key",
	} {
		value, ok := receipt[key]
		if !ok {
			return nil, newInvariantError("receipt_id", "missing field "+key)
		}
		input[key] = value
	}
	if decision, ok := receipt["decision"]; ok && decision != nil {
		input["decision"] = decision
	}
	if observationOutcome, ok := receipt["observation_outcome"]; ok && observationOutcome != nil {
		input["observation_outcome"] = observationOutcome
	}
	if actorChain, ok := receipt["actor_chain"]; ok {
		if values, ok := actorChain.([]any); ok && len(values) > 0 {
			input["actor_chain"] = actorChain
		}
	}
	if evidence, ok := receipt["evidence"]; ok {
		if values, ok := evidence.([]any); ok && len(values) > 0 {
			input["evidence"] = evidence
		}
	}
	if metadata, ok := receipt["metadata"]; ok && metadata != nil {
		input["metadata"] = metadata
	}
	input["trust_level"] = receipt["trust_level"]
	if tenantID, ok := receipt["tenant_id"]; ok && tenantID != nil {
		input["tenant_id"] = tenantID
	}
	return input, nil
}

func receiptSemantics(receipt map[string]any) (string, string) {
	receiptKind, _ := receipt["receipt_kind"].(string)
	boundaryClass, _ := receipt["boundary_class"].(string)
	return receiptKind, boundaryClass
}

func resultLabel(receiptKind string, boundaryClass string, decision string) string {
	if receiptKind == "mediated_decision" && boundaryClass == "prevent" && decision == "allow" {
		return "Authorized"
	}
	switch receiptKind {
	case "trace_observation":
		return "Observed"
	case "advisory_evaluation":
		return "Advisory"
	}
	switch decision {
	case "allow":
		return "Allowed"
	case "deny":
		return "Denied"
	case "cancelled":
		return "Cancelled"
	case "none":
		return "Invalid"
	default:
		return "Incomplete"
	}
}

func validObservationOutcome(value any) bool {
	outcome, ok := value.(string)
	if !ok {
		return false
	}
	switch outcome {
	case "observed", "evaluated", "dropped":
		return true
	default:
		return false
	}
}

func semanticallySignable(receipt map[string]any, decision string) bool {
	receiptKind, boundaryClass := receiptSemantics(receipt)
	trustLevel, _ := receipt["trust_level"].(string)
	observationOutcome := receipt["observation_outcome"]
	switch receiptKind {
	case "mediated_decision":
		return boundaryClass == "prevent" &&
			trustLevel == "mediated" &&
			decision != "none" &&
			observationOutcome == nil
	case "trace_observation":
		return boundaryClass == "detect_only" &&
			trustLevel == "verified" &&
			decision == "none" &&
			validObservationOutcome(observationOutcome)
	case "advisory_evaluation":
		return boundaryClass == "advisory_only" &&
			trustLevel == "advisory" &&
			decision == "none" &&
			validObservationOutcome(observationOutcome)
	default:
		return false
	}
}

File: crates/chio-binding-helpers/src/receipt.rs
use chio_core::{chio_receipt_id, ChioReceipt, Decision, PublicKey};
use serde::{Deserialize, Serialize};

use crate::Result;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ReceiptDecisionKind {
    Allow,
    Deny,
    Cancelled,
    Incomplete,
    None,
}

impl From<Option<&Decision>> for ReceiptDecisionKind {
    fn from(value: Option<&Decision>) -> Self {
        match value {
            Some(Decision::Allow) => Self::Allow,
            Some(Decision::Deny { .. }) => Self::Deny,
            Some(Decision::Cancelled { .. }) => Self::Cancelled,
            Some(Decision::Incomplete { .. }) => Self::Incomplete,
            None => Self::None,
        }
    }
}

fn binding_result_label(
    receipt_kind: &str,
    boundary_class: &str,
    decision: ReceiptDecisionKind,
) -> &'static str {
    if receipt_kind == "mediated_decision"
        && boundary_class == "prevent"
        && decision == ReceiptDecisionKind::Allow
    {
        return "Authorized";
    }
    match receipt_kind {
        "trace_observation" => "Observed",
        "advisory_evaluation" => "Advisory",
        _ => match decision {
            ReceiptDecisionKind::Allow => "Allowed",
            ReceiptDecisionKind::Deny => "Denied",
            ReceiptDecisionKind::Cancelled => "Cancelled",
            ReceiptDecisionKind::Incomplete => "Incomplete",
            ReceiptDecisionKind::None => "Invalid",
        },
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ReceiptVerification {
    pub signature_valid: bool,
    pub parameter_hash_valid: bool,
    pub receipt_id_valid: bool,
    pub decision: ReceiptDecisionKind,
    pub receipt_kind: String,
    pub boundary_class: String,
    pub trust_level: String,
    pub result: String,
    pub authorized: bool,
    pub signer_key_hex: String,
    pub signer_trusted: bool,
    pub ok: bool,
}

pub fn parse_receipt_json(input: &str) -> Result<ChioReceipt> {
    Ok(serde_json::from_str(input)?)
}

pub fn receipt_body_canonical_json(receipt: &ChioReceipt) -> Result<String> {
    Ok(chio_core::canonical_json_string(&receipt.body())?)
}

pub fn verify_receipt(receipt: &ChioReceipt) -> Result<ReceiptVerification> {
    verify_receipt_with_trusted_signers(receipt, &[])
}

pub fn verify_receipt_with_trusted_signers(
    receipt: &ChioReceipt,
    trusted_signers: &[PublicKey],
) -> Result<ReceiptVerification> {
    let receipt_id_valid = chio_receipt_id(&receipt.body())? == receipt.id;
    let semantics = receipt.semantic_fields();
    let semantic_authorized = semantics.is_authorized(receipt.decision.as_ref());
    let decision = ReceiptDecisionKind::from(receipt.decision.as_ref());
    let receipt_kind = semantics.receipt_kind.as_str().to_string();
    let boundary_class = semantics.boundary_class.as_str().to_string();
    let signature_valid = receipt.verify_signature()?;
    let parameter_hash_valid = receipt.action.verify_hash()?;
    let signer_trusted = !trusted_signers.is_empty()
        && trusted_signers
            .iter()
            .any(|signer| signer == &receipt.kernel_key);
    let authorized = semantic_authorized
        && signature_valid
        && parameter_hash_valid
        && receipt_id_valid
        && signer_trusted;
    Ok(ReceiptVerification {
        signature_valid,
        parameter_hash_valid,
        receipt_id_valid,
        decision,
        receipt_kind: receipt_kind.clone(),
        boundary_class: boundary_class.clone(),
        trust_level: receipt.trust_level.as_str().to_string(),
        result: binding_result_label(&receipt_kind, &boundary_class, decision).to_string(),
        authorized,
        signer_key_hex: receipt.kernel_key.to_hex(),
        signer_trusted,
        ok: signature_valid && parameter_hash_valid && receipt_id_valid && signer_trusted,
    })
}

pub fn verify_receipt_with_trusted_signer_hex<S: AsRef<str>>(
    receipt: &ChioReceipt,
    trusted_signer_hex: &[S],
) -> Result<ReceiptVerification> {
    let trusted_signers = parse_trusted_signer_hex(trusted_signer_hex)?;
    verify_receipt_with_trusted_signers(receipt, &trusted_signers)
}

pub fn verify_receipt_json(input: &str) -> Result<ReceiptVerification> {
    let receipt = parse_receipt_json(input)?;
    verify_receipt(&receipt)
}

pub fn verify_receipt_json_with_trusted_signer_hex<S: AsRef<str>>(
    input: &str,
    trusted_signer_hex: &[S],
) -> Result<ReceiptVerification> {
    let receipt = parse_receipt_json(input)?;
    verify_receipt_with_trusted_signer_hex(&receipt, trusted_signer_hex)
}

fn parse_trusted_signer_hex<S: AsRef<str>>(trusted_signer_hex: &[S]) -> Result<Vec<PublicKey>> {
    trusted_signer_hex
        .iter()
        .map(|value| PublicKey::from_hex(value.as_ref()).map_err(Into::into))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::{
        verify_receipt, verify_receipt_json_with_trusted_signer_hex,
        verify_receipt_with_trusted_signer_hex, ReceiptDecisionKind,
    };
    use chio_core::{
        sha256_hex, ChioReceipt, ChioReceiptBody, Decision, GuardEvidence, Keypair, ToolCallAction,
    };

    fn sample_receipt() -> crate::Result<ChioReceipt> {
        let seed = [7u8; 32];
        let keypair = Keypair::from_seed(&seed);
        let action = ToolCallAction::from_parameters(serde_json::json!({
            "path": "/workspace/docs/roadmap.md",
            "mode": "read"
        }))?;
        let body = ChioReceiptBody {
            id: "rcpt-bindings-allow".to_string(),
            timestamp: 1710000100,
            capability_id: "cap-bindings-001".to_string(),
            tool_server: "srv-files".to_string(),
            tool_name: "file_read".to_string(),
            action,
            decision: Some(Decision::Allow),
            receipt_kind: chio_core::ReceiptKind::MediatedDecision,
            boundary_class: chio_core::BoundaryClass::Prevent,
            observation_outcome: None,
            tool_origin: chio_core::ToolOrigin::CallerExecuted,
            redaction_mode: chio_core::RedactionMode::None,
            actor_chain: Vec::new(),
            content_hash: sha256_hex(br#"{"ok":true}"#),
            policy_hash: "policy-bindings-v1".to_string(),
            evidence: vec![GuardEvidence {
                guard_name: "ForbiddenPathGuard".to_string(),
                verdict: true,
                details: Some("path allowed".to_string()),
            }],
            metadata: Some(serde_json::json!({
                "surface": "bindings-test"
            })),
            trust_level: chio_core::TrustLevel::default(),
            tenant_id: None,
            kernel_key: keypair.public_key(),
        };
        Ok(ChioReceipt::sign(body, &keypair)?)
    }

    #[test]
    fn verify_valid_receipt() -> crate::Result<()> {
        let receipt = sample_receipt()?;
        let verification = verify_receipt(&receipt)?;
        assert!(verification.signature_valid);
        assert!(verification.parameter_hash_valid);
        assert!(verification.receipt_id_valid);
        assert_eq!(verification.decision, ReceiptDecisionKind::Allow);
        assert_eq!(verification.receipt_kind, "mediated_decision");
        assert_eq!(verification.boundary_class, "prevent");
        assert_eq!(verification.trust_level, "mediated");
        assert_eq!(verification.result, "Authorized");
        assert!(!verification.authorized);
        assert!(!verification.signer_trusted);
        assert!(!verification.ok);
        assert_eq!(verification.signer_key_hex, receipt.kernel_key.to_hex());
        Ok(())
    }

    #[test]
    fn verify_valid_receipt_with_trusted_signer_is_ok() -> crate::Result<()> {
        let receipt = sample_receipt()?;
        let verification = super::verify_receipt_with_trusted_signers(
            &receipt,
            std::slice::from_ref(&receipt.kernel_key),
        )?;
        assert!(verification.signer_trusted);
        assert!(verification.ok);
        assert!(verification.authorized);
        Ok(())
    }

    #[test]
    fn verify_valid_receipt_with_trusted_signer_hex_is_ok() -> crate::Result<()> {
        let receipt = sample_receipt()?;
        let trusted = vec![receipt.kernel_key.to_hex()];
        let verification = verify_receipt_with_trusted_signer_hex(&receipt, &trusted)?;

        assert!(verification.signer_trusted);
        assert!(verification.ok);
        assert!(verification.authorized);
        Ok(())
    }

    #[test]
    fn verify_receipt_json_with_trusted_signer_hex_is_ok() -> crate::Result<()> {
        let receipt = sample_receipt()?;
        let input = serde_json::to_string(&receipt)?;
        let trusted = vec![receipt.kernel_key.to_hex()];
        let verification = verify_receipt_json_with_trusted_signer_hex(&input, &trusted)?;

        assert!(verification.signer_trusted);
        assert!(verification.ok);
        assert!(verification.authorized);
        Ok(())
    }

    #[test]
    fn invalid_trusted_signer_hex_fails_closed() -> crate::Result<()> {
        let receipt = sample_receipt()?;
        let error =
            verify_receipt_with_trusted_signer_hex(&receipt, &["not-a-public-key"]).unwrap_err();

        assert_eq!(error.code(), crate::ErrorCode::InvalidHex);
        Ok(())
    }

    #[test]
    fn verify_receipt_reports_mismatched_content_addressed_id() -> crate::Result<()> {
        let mut receipt = sample_receipt()?;
        receipt.id = "rcpt-symbolic-invalid".to_string();
        let verification = super::verify_receipt_with_trusted_signers(
            &receipt,
            std::slice::from_ref(&receipt.kernel_key),
        )?;
        assert!(!verification.receipt_id_valid);
        assert!(!verification.signature_valid);
        assert!(!verification.ok);
        Ok(())
    }
}
